learn/testing-ballerina-code/mocking.md

```diff
@@ -0,0 +1,418 @@
+    clientEndpoint = test:mock(http:Client);
+
+    // stub the value of the `url` variable to return the specified string
+    test:prepare(clientEndpoint).getMember("url").thenReturn("https://foo.com/");
+
+    // verify if the specified value is set
+    test:assertEquals(clientEndpoint.url, "https://foo.com/");
+}
+```
+
+#### Stubbing to do nothing
+
+If a function has an optional or no return type specified, this function can be mocked to do nothing when writing
+ test cases.
+
+***Example:***
+
+***main.bal***
+
+```ballerina
+import ballerina/email;
+import ballerina/io;
+
+email:SmtpClient smtpClient = new ("localhost", "admin","admin");
+
+// This function sends out emails to specified email addresses and returns an error if sending failed.
+function sendNotification(string[] emailIds) returns error? {
+    email:Email msg = {
+        'from: "builder@abc.com",
+        subject: "Error Alert ...",
+        to: emailIds,
+        body: ""
+    };
+    email:Error? response = smtpClient->send(msg);
+    if (response is error) {
+	io:println("error while sending the email: " + response.message());
+  	return response;
+    }
+}
+```
+***main_test.bal***
+
+This test stubs the behavior of the `send` function to do nothing for testing the `sendNotification` function.
+
+```ballerina
+import ballerina/test;
+import ballerina/email;
+
+@test:Config {}
+function testSendNotification() {
+    string[] emailIds = ["user1@test.com", "user2@test.com"];
+
+    // create a default mock SMTP client and assign it to the `smtpClient` object
+    smtpClient = test:mock(email:SmtpClient);
+
+    // stub to do nothing when the`send` function is invoked
+    test:prepare(smtpClient).when("send").doNothing();
+
+    // invoke the function to test and verify that no error occurred
+    test:assertEquals(sendNotification(emailIds), ());
+}
+```
+
+## Mocking Functions
+
+The Ballerina test framework provides the capability to mock a function. Using the mocking feature, you can easily mock a function in a module that you are testing or a function of an imported module. This feature will help you to test your Ballerina code independently from other modules and functions.
+
+The object specified with the `@test:Mock {}` annotation will be considered as a mock function that gets triggered in place of the real function.
+
+*   ***moduleName : "&lt;moduleName&gt;"*** - (optional) Name of the module in which the function to be mocked resides in. If the function is within the same module, this can be left blank or "." (no module) can be passed. If the function is in a different module but within the same project, just passing the module name will suffice. For functions in completely separate modules, the fully-qualified module name must be passed, which includes the `orgName` and the `version` i.e., `orgName/module:version`. For native functions, the Ballerina module needs to be specified.
```

Comment:
```suggestion
*   ***moduleName : "&lt;moduleName&gt;"*** - (optional) Name of the module in which the function to be mocked resides in. If the function is within the same module, this can be left blank or "." (no module) can be passed. If the function is in a different module but within the same project, just passing the module name will suffice. For functions in completely-separate modules, the fully-qualified module name must be passed, which includes the `orgName` and the `version` i.e., `orgName/module:version`. For native functions, the Ballerina module needs to be specified.
```